fred2/customstringsdlg.cpp

```diff
@@ -0,0 +1,375 @@
+	: CDialog(CustomStringsDlg::IDD, pParent)
+{
+}
+
+void CustomStringsDlg::DoDataExchange(CDataExchange* pDX)
+{
+	CDialog::DoDataExchange(pDX);
+	DDX_Control(pDX, IDC_CUSTOM_DATA_LISTER, m_data_lister);
+}
+
+// see https://stackoverflow.com/questions/17828258/how-to-prevent-mfc-dialog-closing-on-enter-and-escape-keys
+// This is needed because despite the override of the normal OK and Cancel routes, Enter will still be recognized
+BOOL CustomStringsDlg::PreTranslateMessage(MSG* pMsg)
+{
+	if (!m_text_edit_focus && pMsg->message == WM_KEYDOWN) {
+		if (pMsg->wParam == VK_RETURN || pMsg->wParam == VK_ESCAPE) {
+			return TRUE; // Do not process further
+		}
+	}
+	return CDialog::PreTranslateMessage(pMsg);
+}
+
+BEGIN_MESSAGE_MAP(CustomStringsDlg, CDialog)
+ON_BN_CLICKED(IDC_CUSTOM_ADD, OnStringAdd)
+ON_BN_CLICKED(IDC_CUSTOM_REMOVE, OnStringRemove)
+ON_BN_CLICKED(IDC_CUSTOM_UPDATE, OnStringUpdate)
+
+ON_BN_CLICKED(ID_OK, OnButtonOk)
+ON_BN_CLICKED(ID_CANCEL, OnButtonCancel)
+ON_WM_CLOSE()
+
+ON_LBN_SELCHANGE(IDC_CUSTOM_DATA_LISTER, OnListerSelectionChange)
+
+ON_EN_SETFOCUS(IDC_CUSTOM_STRING, OnEnSetFocusEditString)
+ON_EN_KILLFOCUS(IDC_CUSTOM_STRING, OnEnKillFocusEditString)
+END_MESSAGE_MAP()
+
+/////////////////////////////////////////////////////////////////////////////
+// CustomStringsDlg message handlers
+
+BOOL CustomStringsDlg::OnInitDialog()
+{
+	CDialog::OnInitDialog();
+
+	m_text_edit_focus = false;
+
+	// grab the existing list of custom strings and duplicate it. We only update it if the user clicks OK.
+	m_custom_strings = The_mission.custom_strings;
+
+	update_data_lister();
+
+	// Send default name and values into dialog box
+	UpdateData(FALSE);
+
+	return TRUE; // return TRUE unless you set the focus to a control
+				 // EXCEPTION: OCX Property Pages should return FALSE
+}
+
+void CustomStringsDlg::OnButtonOk()
+{
+	// now we set the custom data to our copy
+	The_mission.custom_strings = m_custom_strings;
+
+	CDialog::OnOK();
+}
+
+void CustomStringsDlg::OnButtonCancel()
+{
+	if (query_modified()) {
+		int z = MessageBox("Do you want to keep your changes?", "Close", MB_ICONQUESTION | MB_YESNOCANCEL);
+		if (z == IDCANCEL) {
+			return;
+		}
+
+		if (z == IDYES) {
+			OnButtonOk();
+			return;
+		}
+	}
+
+	CDialog::OnCancel();
+}
+
+// this is called when you hit the enter key
+void CustomStringsDlg::OnOK()
+{
+	// override MFC default behavior and do nothing
+}
+
+// this is called when you hit the escape key
+void CustomStringsDlg::OnCancel()
+{
+	// override MFC default behavior and do nothing
+}
+
+// this is called when you hit the Close button
+void CustomStringsDlg::OnClose()
+{
+	OnButtonCancel();
+}
+
+void CustomStringsDlg::OnListerSelectionChange()
+{
+	const int index = m_data_lister.GetCurSel();
+
+	if (index == LB_ERR) {
+		// TODO: clear key and data text  edit boxes?
+		return;
+	}
+
+	Assert(m_data_lister.GetCount() > 0);
+	Assert(index >= 0 && index < (int)m_custom_strings.size());
+
```

Comment:
Cue Assertion and casts.  There's more than just here, I think, but this was the first I found.